src/Compilers/Core/Portable/DiagnosticAnalyzer/PerTreeOptionsProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Provide options keyed on <see cref="SyntaxTree"/>.
+    /// </summary>
+    public abstract class PerTreeOptionsProvider
+    {
+        /// <summary>
+        /// Get options for a given <paramref name="tree"/>.
+        /// Null if there are no options registered for the given
```

Comment:
Is it a general roslyn convention to return null if there are no options, rather than returning an empty optionset?

Reply:
I waffled on this a bit. On the one hand, it seems convenient to use an empty one. On the other hand, I thought it may be useful for people to know if they're asking for a SyntaxTree that has no registered options.